server_test.go

```diff
@@ -175,3 +176,158 @@ func TestServer_Authoritative(t *testing.T) {
 		t.Fatal("The authoritative flag should be set")
 	}
 }
+
+func TestServer_AddZone_Simple(t *testing.T) {
+	const (
+		initialZoneName    = "initial.example."
+		additionalZoneName = "additional.example."
+		expectedName       = "resolved.example"
+	)
+
+	// create server with initial zone record
+	srv, err := NewServer(map[string]Zone{
+		initialZoneName: Zone{
+			CNAME: expectedName,
+		},
+	}, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer srv.Close()
+
+	// ensure initial zone record resolves correctly
+	resolvedInitialName, err := srv.Resolver().LookupCNAME(context.Background(), initialZoneName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if expectedName != resolvedInitialName {
+		t.Fatalf("expected: %s; got: %s", expectedName, resolvedInitialName)
+	}
+
+	// add additional zone record
+	err = srv.AddZone(additionalZoneName, Zone{
+		CNAME: expectedName,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// ensure additional zone record resolves correctly
+	resolvedAdditionalName, err := srv.Resolver().LookupCNAME(context.Background(), additionalZoneName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if expectedName != resolvedAdditionalName {
+		t.Fatalf("expected: %s; got: %s", expectedName, resolvedInitialName)
+	}
+}
+
+func TestServer_AddZone_Existing(t *testing.T) {
+	const (
+		initialZoneName = "initial.example."
+		expectedName    = "expected.example"
+		unexpectedName  = "unexpected.example"
+	)
+
+	var expectedErr = fmt.Errorf(ErrExistingZoneFmt, initialZoneName)
+
+	// create server with initial zone record
+	srv, err := NewServer(map[string]Zone{
+		initialZoneName: Zone{
+			CNAME: expectedName,
+		},
+	}, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer srv.Close()
+
+	// ensure initial zone record resolves correctly
+	resolvedInitialName, err := srv.Resolver().LookupCNAME(context.Background(), initialZoneName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if expectedName != resolvedInitialName {
+		t.Fatalf("expected: %q but got: %q", initialZoneName, resolvedInitialName)
+	}
+
+	// attempt to add existing zone record
+	err = srv.AddZone(initialZoneName, Zone{
+		CNAME: unexpectedName,
+	})
+	if expectedErr.Error() != err.Error() {
```

Comment:
Reading this code makes me really appreciate the testify package.

Reply:
:100: Agreed!